File: momose/src/core/SimulationManager.java
package core;

import java.io.File;
import java.util.Iterator;
import java.util.Vector;


import javax.swing.JFileChooser;
import javax.swing.JOptionPane;

import parsers.ConfigSimParser;
import utils.MomoseFileFilter;
import utils.Preferences;
import utils.Utils;
import datarecorders.btgRecorder.BTGRecorderBuilder;
import datarecorders.debugRecorder.DebugRecorderBuilder;
import datarecorders.ecraRecorder.ECRARecorderBuilder;
import datarecorders.ns2MobilityRecorder.NS2MobilityRecorderBuilder;
import datarecorders.viewerRecorder.ViewerRecorderBuilder;
import datarecorders.scramRecorder.ScramRecorderBuilder;

import engine.Scenario;
import engine.SimTime;

import models.ecraModel.EcraModelBuilder;
import models.hotSpotModel.HotSpotModelBuilder;
import models.nomadicModel.NomadicModelBuilder;
import models.pursueModel.PursueModelBuilder;
import models.randomWalkModel.RandomWalkModelBuilder;
import models.randomWaypointModel.RandomWaypointModelBuilder;
import models.rwpTouristModel.RWPTouristModelBuilder;
import models.touristGroupModel.TouristGroupModelBuilder;

import gui.ConfigSimulationWnd;
import gui.Momose;
import gui.SimulationWnd;


public class SimulationManager
 {
  private static int FROM_WND=0;
  private static int FROM_FILE=2;
 
  private Momose moveWnd;
  private int wndOpen;
  private int simID;
    
  public SimulationManager(Momose moveWnd)
   {
 this.moveWnd=moveWnd;
 wndOpen=0;
 simID=0;
   }//Fine costruttore
  
  public Momose getMoveWnd()
   {return moveWnd;}
  
  private void incrCounters()
   { simID++; wndOpen++; }
  
  public void wndClosed()
   { wndOpen--; }
   
 synchronized public void newConfigSimWnd() 
  {
   incrCounters();  
   //Creo  la finestra di configurazione della simulazione  
   ConfigSimulationWnd configSimWnd=new ConfigSimulationWnd(this,simID,true,
                               getModelBuilders(),getRecorderBuilders());
   //Mostro la finestra
   showConfigWnd(configSimWnd);
  }//Fine showConfigSim
 
 
 
 private void showConfigWnd(ConfigSimulationWnd cfgSimWnd)
  {
   //Aggancio la dlg alla finestra padre
   moveWnd.getDesktopPane().add(cfgSimWnd);
   //Mostro la finestra
   cfgSimWnd.setVisible(true);
  }//Fine showConfigWnd
 
 public File showLoadFileDlg(String curDir,String type)
  {
  
   JFileChooser fc=new JFileChooser();
   //Setto la directory corrente
   fc.setCurrentDirectory(new File(curDir));
   //Setto il filtro
   fc.setFileFilter(new MomoseFileFilter(type));
   //Mostro la finestra
   int returnVal=fc.showOpenDialog(moveWnd);
   
   if(returnVal==JFileChooser.APPROVE_OPTION)
    return fc.getSelectedFile();
   else
 return null;  
  }//Fine showLoadFileDlg
 
 public void LoadSimFromFile()
  {
   //Mostro la finestra di caricamento
   Preferences pref=Utils.getPreferences(Momose.PREF_FILE);  
   File file=showLoadFileDlg(pref.getConfigDir(),"config");
   if(file!=null)
    {    
  //Creo e mostro la finestra di simulazione  
     SimulationWnd simWnd=createSimFromFile(file);
     if(simWnd!=null)
      {
       //Incremento il conteggio delle finestra aperte  
       incrCounters();
       //Mostro la finestra di simulazione
       showSimWnd(simWnd,simID);
      } 
    }     
  }//Fine loadSimFromFile
 
 public void loadSimFromWnd(ConfigSimulationWnd cfgWnd)
  {
   SimulationWnd simWnd=createSimFromWnd(cfgWnd);
   if(simWnd!=null)
    {
     //Disabilito i componenti   
  cfgWnd.enableComp(false);  
  //Incremento il numero di finestre aperte   
  wndOpen++;  
     //Mostro la finestra di simulazione
     showSimWnd(simWnd,cfgWnd.getSimID());
     
    }   
  }//Fine loadSimFromWnd
 
 private void showSimWnd(SimulationWnd simWnd,int simID)
  {
   //Sett l'ID della simualazione  
   simWnd.setSimID(simID);
   //Aggancio la dlg alla finestra padre
   moveWnd.getDesktopPane().add(simWnd);
   //Mostro la finestra
   simWnd.setVisible(true);  
  }//Fine showSimWnd
 
 private SimulationWnd createSimFromWnd(ConfigSimulationWnd cfgWnd)
  {
   //Creo il tempo
   SimTime simTime=new SimTime(cfgWnd.getT(),cfgWnd.getDT(),cfgWnd.getLOOPS());
   if(ctrlSimTime(simTime)==false)
    return null;
   
   //Creo lo scenario
   Scenario scenario=SimLoader.loadScenarioFromCnfWnd(cfgWnd);
   if(scenario==null)
  {
   JOptionPane.showMessageDialog(moveWnd,"Unable to open scenario file...\n" +
                                 "Loading aborted!",
                                 "Error",JOptionPane.ERROR_MESSAGE);    
   return null;
  }  
 
   //Prendo i modelli
   Vector modBuilders=prepareModBuilders(null,FROM_WND,cfgWnd.getModBuilders(),cfgWnd.getModBuildNamesSel());
   if(modBuilders==null)
  return null;
   
   //Prendo i recorders
   Vector recBuilders=prepareRecBuilders(null,FROM_WND,cfgWnd.getRecBuilders(),cfgWnd.getRecBuildNamesSel());
   if(recBuilders==null)
   return null;
   
   //Creo e ritorno la finestra di simulazione
   return new SimulationWnd(simTime,scenario,modBuilders,recBuilders,FROM_WND,this,cfgWnd);
  }//Fine createSimFromDlg
 
 private boolean ctrlSimTime(SimTime simTime) 
  {
   if((simTime.getN()<=0)||(simTime.getDT()<=0)||(simTime.getT()<=0))
    {
  JOptionPane.showMessageDialog(moveWnd,"Error in setup of simulation time...\n"+
                                   "Loading aborted!",
                                   "Error",JOptionPane.ERROR_MESSAGE);  
  return false;
    } 
 else
  return true;
  }//Fine ctrlSimTime

private SimulationWnd createSimFromFile(File cfgFile) 
  {
   //Leggo il file di configurazione  
   ConfigSimParser cfgParser=parseConfigFile(cfgFile);
   if(cfgParser==null)
 return null;
 
   //Prelevo il tempo
   SimTime simTime=getSimTimeFromParser(cfgParser);
   if(simTime==null)
 return null;
   //Controllo il formato del tempo
   if(ctrlSimTime(simTime)==false)
  return null;  
   
   //Prelevo lo scenario
   Scenario scenario=getScenarioFromParser(cfgParser);
   if(scenario==null)
  return null; 
      
   //Prendo i modelli
   Vector modBuilders=prepareModBuilders(cfgFile,FROM_FILE,getModelBuilders(),cfgParser.getModBuildNames());
   if(modBuilders==null)
  return null;
   
   //Prendo i recorders
   Vector recBuilders=prepareRecBuilders(cfgFile,FROM_FILE,getRecorderBuilders(),cfgParser.getRecBuildNames());
   if(recBuilders==null)
   return null;
   
   //Creo e ritorno la finestra di simulazione
   return new SimulationWnd(simTime,scenario,modBuilders,recBuilders,FROM_FILE,this,null);
  }//Fine createSimulation
 
 private Vector prepareModBuilders(File cfgFile,int loadType,Vector modelsBuilders,Vector modBuildersNames)
  {
   Vector modBuildGen=setupBuilders(cfgFile,loadType,modelsBuilders,modBuildersNames);
   if(modBuildGen==null)
     { 
      JOptionPane.showMessageDialog(moveWnd,"Error in setup of models builders...\n"+
                                    "Loading aborted!",
                                    "Error",JOptionPane.ERROR_MESSAGE);  
      return null;
     }
    return modBuildGen;
   }//Fine getModelsBuilders  
 
 private Vector prepareRecBuilders(File cfgFile,int loadType,Vector recsBuilders,Vector recBuildersNames)
  {
   Vector recBuilders=setupBuilders(cfgFile,loadType,recsBuilders,recBuildersNames);
   if(recBuilders==null)
    { 
     JOptionPane.showMessageDialog(moveWnd,"Error in setup of data-recordes builders...\n"+
                                   "Loading aborted!",
                                   "Error",JOptionPane.ERROR_MESSAGE); 
     return null;
    }  
   return recBuilders;
  }//Fine getRecBuilders
 
 private ConfigSimParser parseConfigFile(File cfgFile)
  {
   ConfigSimParser cnfSimParser=SimLoader.loadDataFromCnfFile(cfgFile);
   if(cnfSimParser==null)
 {  
  JOptionPane.showMessageDialog(moveWnd,"Error in parsing main config-file...\n" +
                                  "Loading aborted!",
                                "Error",JOptionPane.ERROR_MESSAGE);  
      return null;
    }
   else
    {
     if(cnfSimParser.isConfigFile())
      { return cnfSimParser; }
     else
      {
       JOptionPane.showMessageDialog(moveWnd,"The input file is not a config-file...\n"+
                                     "Loading aborted!" ,
                                     "Error",JOptionPane.ERROR_MESSAGE);
       return null;
      }
    } 
  }//Fine loadDataInParser 
 
 private SimTime getSimTimeFromParser(ConfigSimParser cnfSimParser)
  {
   SimTime simTime=cnfSimParser.getSimTime();
   if(simTime==null)
 {
  JOptionPane.showMessageDialog(moveWnd,"Error in parsing sim-time data...\n"+
                                        "Loading aborted!", 
                                        "Error",JOptionPane.ERROR_MESSAGE);  
   return null;
    }
   return simTime;
  }//Fine getSimTimeFormParser
 
 private Scenario getScenarioFromParser(ConfigSimParser cnfSimParser)
  {
   Scenario scenario=cnfSimParser.getScenario();
   if(scenario==null)
  {
   JOptionPane.showMessageDialog(moveWnd,"Error in parsing scenario data...\n" +
                                 "Loading aborted!",
                                         "Error",JOptionPane.ERROR_MESSAGE);  
   return null;
  } 
   return scenario;
  }//Fine getScenarioFromParser
 
 private Vector setupBuilders(File cfgFile,int loadType,Vector vecTarget,Vector modNames) 
  {
   //Creo i builder
   Vector builders=getBuildersVector(vecTarget,modNames);
   if(builders==null)
  return null;
   
   //Se carico da file configuro il builder 
   if(loadType==FROM_FILE)
    {  
     //Faccio il parsing dei file di configurazione dei builder dei modelli
     boolean opOk=SimLoader.loadBuildersConfigData(builders,cfgFile);
     if(opOk==false)
   return null;
    }  
   //ritorno i builder
   return builders;
  }//Fine createModels
 
 private Vector getBuildersVector(Vector buildersSelected,Vector names)
  {
   Vector builders=new Vector();
   Iterator i=names.iterator();
   while(i.hasNext())
    {
  //Estraggo il nome 
  String nextName=(String)i.next();  
  Builder builder=getBuilder(buildersSelected,nextName);
  //Se il builder esiste lo metto nel vettore
  if(builder!=null)
   {builders.add(builder);} 
  else
   {
       //Se non trovo il builder mostro messaggio di errore
    JOptionPane.showMessageDialog(moveWnd,nextName+" not found in current system...",
                              "Warning",JOptionPane.WARNING_MESSAGE);  
    return null;
  } 
 }    
   return builders;  
 }//Fine createBuildersVector
 
 private Builder getBuilder(Vector builders,String name)
  {
   String builderName=null; 
   for(int i=0;i<builders.size();i++)
 {
  Builder nextBuilder=(Builder)builders.elementAt(i);
  builderName=nextBuilder.getName();
  if(builderName.equals(name))
    return nextBuilder;
 }   
  return null;
 }//Fine getModelBuilder
 
 //////////////
 synchronized public void loadConfigSimWnd() 
  {
   //Mostro la finestra per aprire il file di configurazione
   Preferences pref=Utils.getPreferences(Momose.PREF_FILE);    
   File file=showLoadFileDlg(pref.getConfigDir(),"config");
   if(file!=null)
    {   
  ConfigSimulationWnd cfgWnd=createCfgWndFromFile(file);
  if(cfgWnd!=null)
   showConfigWnd(cfgWnd);
 }
  }//Fine loadConfigWnd
 
 private ConfigSimulationWnd createCfgWndFromFile(File cfgFile)
  {
   //Incremento i contatori  
   incrCounters();  
   //Creo  la finestra di configurazione della simulazione
   ConfigSimulationWnd cfgWnd=new ConfigSimulationWnd(this,simID,false,
                              getModelBuilders(),getRecorderBuilders());  
     
   //Leggo il file di configurazione  
   ConfigSimParser cfgParser=parseConfigFile(cfgFile);
   if(cfgParser==null)
 return null;
 
   //Prelevo il tempo
   SimTime simTime=getSimTimeFromParser(cfgParser);
   if(simTime!=null)
    {
  //Controllo la correttezza 
  if(ctrlSimTime(simTime)==false)
   return null;
  else
   cfgWnd.setTimeData(simTime);
    }
   else
 return null; 
   
   //Prelevo lo scenario
   Scenario scenario=getScenarioFromParser(cfgParser);
   if(scenario!=null)
 cfgWnd.setScenarioCfg(scenario);
   else
 return null; 
   
   //Setto i ModelBuilder
   Vector modBuilders=prepareModBuilders(cfgFile,FROM_FILE,cfgWnd.getModBuilders(),
                                      cfgParser.getModBuildNames());
   if(modBuilders==null)
   return null; 
   //Setto le finestre di dialogo
   setDlgData(modBuilders);
   //Seleziono i builder nella lista dei modelli della finestra
   cfgWnd.setBuildersChecked(modBuilders,ConfigSimulationWnd.MODELS);
   
   
   //Prendo i recorders
   Vector recBuilders=prepareRecBuilders(cfgFile,FROM_FILE,cfgWnd.getRecBuilders(),
                               cfgParser.getRecBuildNames());
   if(recBuilders==null)
   return null;
   
   //Setto le finestre di dialogo
   setDlgData(recBuilders);
   //Seleziono i builder nella lista dei modelli dei recorders 
   cfgWnd.setBuildersChecked(recBuilders,ConfigSimulationWnd.RECORDERS);
   
   //Ritorno la finestra di configurazione
   return cfgWnd;  
  }//Fine createCfgWndFromFile
 
 private void setDlgData(Vector builders)
  {
   Iterator i=builders.iterator();
   while(i.hasNext())
 {
  Builder nextBuilder=(Builder)i.next();
  //Setto la finestra di configurazione di questo builder
  nextBuilder.setConfigDlgFromFile(); 
 }        
   }//Fine setDlgData
 
////////////////////////////////////
  private Vector getModelBuilders()
   {
 Vector modelsBuilders=new Vector();  
 
 //Random Walk Model
 modelsBuilders.add(new RandomWalkModelBuilder("Random walk Model"));
 modelsBuilders.add(new RandomWaypointModelBuilder("Random waypoint Model")); 
 modelsBuilders.add(new EcraModelBuilder("ECRA Model")); 
 modelsBuilders.add(new HotSpotModelBuilder("Hot-Spot Model")); 
 modelsBuilders.add(new NomadicModelBuilder("Nomadic Model")); 
 modelsBuilders.add(new PursueModelBuilder("Pursue Model")); 
 modelsBuilders.add(new TouristGroupModelBuilder("Tourist group Model"));
 modelsBuilders.add(new RWPTouristModelBuilder("RWP tourist Model")); 
 //Insert here new ModelBuilder
 //modelsBuilder.add(new NewModelBuilder());
 
 return modelsBuilders;
   }
  
  private Vector getRecorderBuilders()
   {
 Vector recordersBuilders=new Vector();    
 
 //ViewerRecorder
    recordersBuilders.add(new ViewerRecorderBuilder("Viewer recorder"));
    recordersBuilders.add(new ScramRecorderBuilder("Scram recorder"));
    recordersBuilders.add(new DebugRecorderBuilder("Debug recorder"));
    recordersBuilders.add(new NS2MobilityRecorderBuilder("ns-2 recorder"));
    recordersBuilders.add(new BTGRecorderBuilder("Bluetooh graph recorder"));
    recordersBuilders.add(new ECRARecorderBuilder("ECRA recorder")); 
    //Insert here new DataRecorderBuilder
 //recoredersBuilder.add(new NewDataRecorderBuilder());  
    
    return recordersBuilders;
   }
 }//Fine simulation manager

File: momosecpp/src/SimulationManager.cpp
#include"SimulationManager.h"


#include"Utils.h"


#include"ViewerRecorderBuilder.h"
#include"ViewerRecorder.h"
#include"BaseParser.h"


#include"RandomWalkModelBuilder.h"
#include"RandomWaypointModelBuilder.h"
#include"PursueModelBuilder.h"
#include"NomadicModelBuilder.h"
#include"EraModelBuilder.h"
#include"HotSpotModelBuilder.h"

#define BUFSIZE 1024

#ifdef WINDOWS
  #define _SEPARATOR_ '\'
#else
  #define _SEPARATOR_ '/'
#endif  

SimulationManager::SimulationManager() 
 {
   modelsBuilders=new vector<ModelBuilder*>();    
   recordersBuilders=new vector<DataRecorderBuilder*>();
    
   //Carico il vettore dei modelli	 
   setModelBuilders();
   //Carico il vettore dei data-recorder
   setRecorderBuilders();	 
 }//Fine costruttore
	
SimulationManager::~SimulationManager() 
 {
   //delete(&modelsBuilders);	 
   //delete(&recordersBuilders);
 }//Fine distruttore	
	
	
Simulation* SimulationManager::createSimulation(string filename,ostream *outStream,bool verbose,bool progress)  
  {
     char *filenameChar=Utils::convertStrToCharVect(filename);	  
	  
     //Leggo il file di configurazione	  
     ConfigSimParser	 *configSimParser=parseConfigFile(filenameChar);	 
	  
     if(configSimParser==NULL)
     {
       cerr<<"ERROR: Main config-file parsing failed... Loading aborted!"<<endl;
       exit(1);		     
     }	     
     
     //Prelevo il tempo	   
     SimTime *simTime=configSimParser->getSimTime();
     if(simTime==NULL)
     {
       cerr<<"ERROR: Simulation time data parsing failed... Loading aborted!"<<endl;
       exit(1);	
     }	     
     //Controllo il formato del tempo
     ctrlSimTime(simTime);
     
     //Prelevo lo scenario
     Scenario *scenario=configSimParser->getScenario();  
      if(scenario==NULL)
      {
        cerr<<"ERROR: Scenario data parsing failed... Loading aborted!"<<endl;
        exit(1);	
      }	     
     
      //Creo i modelli
      vector<ModelBuilder*> *modBuilders=setupModBuilders(configSimParser->getModBuildNames(),filename);
      vector<Model*> *models=getModels(modBuilders);
      
      //Creo i modelli       
      vector<DataRecorderBuilder*> *recBuilders=setupRecBuilders(configSimParser->getRecBuildNames(),filename);
      vector<DataRecorder*> *recorders=getRecorders(recBuilders);
      
      //Creo e ritorno la simulazione
      Simulation *simulation=new Simulation(scenario,models,recorders,simTime,outStream,verbose,progress);      
      return simulation;	  
  }//Fine createSimulation
  
ConfigSimParser* SimulationManager::parseConfigFile(const char* filename)
   {
     //Faccio il parsing del file	   
     ConfigSimParser	  *configSimParser=new ConfigSimParser();
     int res=parseXMLFile(filename,configSimParser);	   
	   
     if(res==-1)   
       {
	 cerr<<"ERROR: Main config-file parsing failed... Loading aborted!"<<endl;
         exit(1);	
        }     
      else
       {
	 if(!configSimParser->isConfigFile())
	  {
	    cerr<<"ERROR: The input file is not a config-file... Loading aborted!"<<endl;
	    exit(1);		  
	  }		 
	}      
	   
     return configSimParser;  
   }//Fine parseConfigFile  
  
int SimulationManager::parseXMLFile(const char* filename, XMLEvtHandler* pHandler)  
  {
    // Creo il buffer su cui inserire i dati letti dal file XML
    char buf[BUFSIZE];
		
    //Inizializzo l'istanza del parser
    XMLParser parser(pHandler);
    XML_Char sep;
    parser.Create("UTF-8",&sep);
    parser.EnableElementHandler();
    parser.EnableCharacterDataHandler();

    //Apro il file xml 
    FILE*  fh=fopen(filename,"r");
    if(fh==NULL)
	{return -1;} 	 
         	    
    //Faccio il parsing del file aperto    
    bool atEnd=false;
    size_t len;
    int done;
    do 
	{
          len=fread(buf, 1, sizeof(buf), fh);
          done=len<sizeof(buf);
          parser.Parse(buf,len,done);
	} 
    while (!done);
	
   //Chiudo il file xml	
   fclose(fh);   
    
   return 0;   
 }//Fine parseXMLFile 
 
bool SimulationManager::ctrlSimTime(SimTime *simTime)
  {
    if((simTime->getN()<=0)||(simTime->getDT()<=0)||(simTime->getT()<=0))
    {
      cerr<<"ERROR: Bad format of simulation time... Loading aborted!"<<endl;
      exit(1);                             
    }	
  }//Fine ctrlSimTime 
  

vector<ModelBuilder*>* SimulationManager::setupModBuilders(vector<string> names,string cfgFileName)
  {
    //Carico i builders	  
    vector<ModelBuilder*> *builders=getModBuildersVector(names);
	  
     vector<ModelBuilder*>::iterator i;
     for(i=builders->begin();i!=builders->end();i++)
       { setupBuilder(*i,cfgFileName); }  
       
    return builders;		  
  }//Fine   setupModBuilders
  
  
vector<DataRecorderBuilder*>* SimulationManager::setupRecBuilders(vector<string> names,string cfgFileName)
  {
    //Carico i builders	  
    vector<DataRecorderBuilder*> *builders=getRecBuildersVector(names);
	  
    vector<DataRecorderBuilder*>::iterator i;
     for(i=builders->begin();i!=builders->end();i++)
       { setupBuilder(*i,cfgFileName);  }  
       
    return builders;		  
  }//Fine   setupRecBuilders  
  
void SimulationManager::setupBuilder(Builder *builder,string cfgFileName)  
   {
     BaseParser *parser=builder->getConfigParser();
     if(parser!=NULL)
	 {
	    string configStr=getCfgFilePath(builder->getName(),cfgFileName);	 
            int res=parseXMLFile(Utils::convertStrToCharVect(configStr),parser);
            if(res==-1)
	     {
	       cerr<<"ERROR: Parsing " <<configStr<<" failed... Loading aborted!"<<endl;
               exit(1);                        
	     }		    
         }			   
   }//Fine setupBuilder
  
  
string SimulationManager::getCfgFilePath(string modelName,string mainCfgFile)
    {
      //Cerco l'inizio del nome del file di configurazione	    
      int index=mainCfgFile.size();	     
      char curCar;
      do
       {
	 index--;        
	 if(index>=0)      
	  curCar=mainCfgFile[index];
	}
     while((curCar!=_SEPARATOR_)&&(index>=0));	
        
	//Inserisco il nome del modello
	mainCfgFile.insert(index+1,modelName+"-");
	    
     //Ritorno il nuovo nome del file	
     return mainCfgFile; 	
   }//Fine getCfgFilePath
  
  
  
vector<ModelBuilder*>* SimulationManager::getModBuildersVector(vector<string> names)
  {
     ModelBuilder* modBuilder=NULL;	  
     vector<ModelBuilder*> *builders=new vector<ModelBuilder*>();
	  
     vector<string>::iterator i;
     for(i=names.begin();i!=names.end();i++)
       {
	 modBuilder=searchModName(*i);
	 if(modBuilder!=NULL) 
	   builders->push_back(modBuilder);
         else 
	  { cerr<<"WARNING: "<<*i<<" not found in current system!"<<endl; }	 
	}
    return builders;	  
  }//Fine  getBuildersVector 
  
  
vector<DataRecorderBuilder*>* SimulationManager::getRecBuildersVector(vector<string> names)
  {
     DataRecorderBuilder* recBuilder=NULL;	  
     vector<DataRecorderBuilder*> *builders=new vector<DataRecorderBuilder*>();
	  
     vector<string>::iterator i;
     for(i=names.begin();i!=names.end();i++)
       {
	recBuilder=searchRecName(*i);
	 if(recBuilder!=NULL) 
	   builders->push_back(recBuilder);
         else 
	  { cerr<<"WARNING: "<<*i<<" not found in current system!"<<endl; }	 
	}
    return builders;	  
  }//Fine  getBuildersVector   
  
 ModelBuilder* SimulationManager::searchModName(string name)
   {
     vector<ModelBuilder*>::iterator j;
     for(j=modelsBuilders->begin();j!=modelsBuilders->end();j++)
       {
	 if(name==(*j)->getName())
 	   return *j; 	 
       }	 
     return NULL; 
   }//Fine searchName  
   
 DataRecorderBuilder* SimulationManager::searchRecName(string name)
   {
     vector<DataRecorderBuilder*>::iterator j;
     for(j=recordersBuilders->begin();j!=recordersBuilders->end();j++)
       {
	 if(name==(*j)->getName())
 	   return *j; 	 
       }	 
     return NULL; 
   }//Fine searchName   
   
vector<Model*>* SimulationManager::getModels(vector<ModelBuilder*> *builders)
   {
     vector<Model*> *models=new vector<Model*>();
     vector<ModelBuilder*>::iterator i;
     for(i=builders->begin();i!=builders->end();i++)
       { models->push_back((*i)->create());}  	   
       
      return models; 
   }//Fine getModels  

vector<DataRecorder*>* SimulationManager::getRecorders(vector<DataRecorderBuilder*> *builders)
   {
     vector<DataRecorder*> *recorders=new vector<DataRecorder*>();
     vector<DataRecorderBuilder*>::iterator i;
     for(i=builders->begin();i!=builders->end();i++)
       { recorders->push_back((*i)->create());}  	   
       
      return recorders; 
   }//Fine getModels     
   
 
  
////////////////////////////////////////////////////////////////////  
void SimulationManager::setModelBuilders()
   {
     modelsBuilders->push_back(new RandomWalkModelBuilder("Random walk Model"));
     modelsBuilders->push_back(new RandomWaypointModelBuilder("Random waypoint Model")); 
     modelsBuilders->push_back(new EraModelBuilder("ERA Model")); 
     modelsBuilders->push_back(new HotSpotModelBuilder("Hot-Spot Model")); 
     modelsBuilders->push_back(new NomadicModelBuilder("Nomadic Model")); 
     modelsBuilders->push_back(new PursueModelBuilder("Pursue Model")); 
     //Insert here new ModelBuilder
     //modelsBuilder->push_back(new NewModelBuilder());
  }//end getModelBuilders
  
  void SimulationManager::setRecorderBuilders()
   {
     recordersBuilders->push_back(new ViewerRecorderBuilder("Viewer recorder"));  
     //recordersBuilders->push_back(new DebugRecorderBuilder("Debug recorder")); 
     //Insert here new DataRecorderBuilder
	//recoredersBuilder->push_back(new NewDataRecorderBuilder());  
   }//End getRecorderBuilder  
	 